File: sdk/python/examples/langchain_integration.py
"""
ARKVOID SDK – LangChain Integration Example (Python)

Custom BaseCallbackHandler that sends every LLM call,
chain run, and tool invocation to ARKVOID automatically.

pip install arkvoid langchain langchain-openai
"""

import os
import time
from typing import Any, Dict, List, Optional, Union
from uuid import UUID

from langchain.callbacks.base import BaseCallbackHandler
from langchain.schema import LLMResult
from langchain_openai import ChatOpenAI
from langchain.agents import AgentExecutor, create_openai_tools_agent
from langchain_core.tools import tool
from langchain import hub

from arkvoid import ArkvoidClient, ToolCallRecord

# ─────────────────────────────────────────────
# ARKVOID LangChain Callback Handler
# ─────────────────────────────────────────────

class ArkvoidCallbackHandler(BaseCallbackHandler):
    """
    LangChain callback handler that sends all LLM + tool events to ARKVOID.

    Usage:
        handler = ArkvoidCallbackHandler(client=arkvoid, agent_slug="my-agent")
        llm = ChatOpenAI(model="gpt-4o", callbacks=[handler])
    """

    def __init__(self, client: ArkvoidClient, agent_slug: str):
        super().__init__()
        self.client = client
        self.agent_slug = agent_slug
        self._run_start: Dict[str, float] = {}
        self._pending_tools: Dict[str, List[ToolCallRecord]] = {}

    # ── LLM events ─────────────────────────────

    def on_llm_start(
        self,
        serialized: Dict[str, Any],
        prompts: List[str],
        *,
        run_id: UUID,
        **kwargs: Any,
    ) -> None:
        run_key = str(run_id)
        self._run_start[run_key] = time.time()
        self._pending_tools[run_key] = []

    def on_llm_end(
        self,
        response: LLMResult,
        *,
        run_id: UUID,
        **kwargs: Any,
    ) -> None:
        run_key = str(run_id)
        start = self._run_start.pop(run_key, None)
        duration_ms = int((time.time() - start) * 1000) if start else None
        tool_calls = self._pending_tools.pop(run_key, [])

        llm_output = response.llm_output or {}
        model_name = (
            llm_output.get("model_name")
            or llm_output.get("model")
            or "unknown"
        )
        usage = llm_output.get("token_usage", {})

        self.client.trace(
            action="llm_call",
            risk_level="low",
            agent=self.agent_slug,
            model_provider="openai",
            model_name=model_name,
            input_tokens=usage.get("prompt_tokens"),
            output_tokens=usage.get("completion_tokens"),
            duration_ms=duration_ms,
            tool_calls=tool_calls or None,
            metadata={
                "source": "langchain",
                "run_id": run_key,
                "generation_count": len(response.generations),
            },
            tags=["langchain", "llm"],
        )

    def on_llm_error(
        self,
        error: Union[Exception, KeyboardInterrupt],
        *,
        run_id: UUID,
        **kwargs: Any,
    ) -> None:
        run_key = str(run_id)
        start = self._run_start.pop(run_key, None)
        self._pending_tools.pop(run_key, None)

        self.client.trace(
            action="llm_error",
            risk_level="high",
            agent=self.agent_slug,
            duration_ms=int((time.time() - start) * 1000) if start else None,
            metadata={
                "source": "langchain",
                "run_id": run_key,
                "error": str(error),
                "success": False,
            },
            tags=["langchain", "error"],
        )

    # ── Tool events ─────────────────────────────

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: str,
        *,
        run_id: UUID,
        parent_run_id: Optional[UUID] = None,
        **kwargs: Any,
    ) -> None:
        self._run_start[str(run_id)] = time.time()
        tool_name = serialized.get("name", "unknown_tool")

        if parent_run_id:
            parent_key = str(parent_run_id)
            if parent_key in self._pending_tools:
                calls = self._pending_tools[parent_key]
                calls.append(
                    ToolCallRecord(
                        tool_name=tool_name,
                        call_index=len(calls),
                        input={"query": input_str},
                        status="success",
                    )
                )

    def on_tool_end(
        self,
        output: str,
        *,
        run_id: UUID,
        parent_run_id: Optional[UUID] = None,
        **kwargs: Any,
    ) -> None:
        run_key = str(run_id)
        start = self._run_start.pop(run_key, None)
        latency_ms = int((time.time() - start) * 1000) if start else None

        if parent_run_id:
            parent_key = str(parent_run_id)
            calls = self._pending_tools.get(parent_key, [])
            if calls:
                last = calls[-1]
                last.latency_ms = latency_ms
                last.output = {"result": output[:500]}

    def on_tool_error(
        self,
        error: Union[Exception, KeyboardInterrupt],
        *,
        run_id: UUID,
        **kwargs: Any,
    ) -> None:
        run_key = str(run_id)
        start = self._run_start.pop(run_key, None)

        self.client.trace(
            action="tool_error",
            risk_level="medium",
            agent=self.agent_slug,
            duration_ms=int((time.time() - start) * 1000) if start else None,
            metadata={"error": str(error), "run_id": run_key},
            tags=["langchain", "tool", "error"],
        )


# ─────────────────────────────────────────────
# Usage
# ─────────────────────────────────────────────

arkvoid = ArkvoidClient(
    api_key=os.environ["ARKVOID_API_KEY"],
    agent="research-agent",
)

handler = ArkvoidCallbackHandler(client=arkvoid, agent_slug="research-agent")

llm = ChatOpenAI(model="gpt-4o", temperature=0, callbacks=[handler])


# ── Simple chain ─────────────────────────────

def simple_chain(prompt: str) -> str:
    """LLM call with auto-tracing via callback."""
    result = llm.invoke(prompt)
    return result.content


# ── Agent with tools ─────────────────────────

@tool
def web_search(query: str) -> str:
    """Search the web for current information."""
    # Simulated — swap with real search API
    return f"[Simulated search results for: {query}]"


@tool
def calculator(expression: str) -> str:
    """Evaluate a mathematical expression."""
    try:
        return str(eval(expression, {"__builtins__": {}}, {}))
    except Exception as e:
        return f"Error: {e}"


async def run_agent(task: str) -> str:
    """Run a LangChain agent with full ARKVOID tracing."""
    tools = [web_search, calculator]
    prompt = hub.pull("hwchase17/openai-tools-agent")

    agent = create_openai_tools_agent(llm=llm, tools=tools, prompt=prompt)
    executor = AgentExecutor(agent=agent, tools=tools, callbacks=[handler])

    result = await executor.ainvoke({"input": task})
    return result["output"]


# ─────────────────────────────────────────────
# Run
# ─────────────────────────────────────────────

if __name__ == "__main__":
    import asyncio

    print("🔗 ARKVOID LangChain Integration (Python)\n")

    answer = simple_chain("What is the difference between RAG and fine-tuning?")
    print(f"Answer: {answer[:150]}...\n")

    result = asyncio.run(run_agent("Calculate 15% of 8500 and search for AI safety news"))
    print(f"Agent: {result[:200]}")

File: sdk/javascript/examples/langchain-integration.ts
/**
 * ARKVOID SDK – LangChain Integration Example
 *
 * Integrates ARKVOID with LangChain via a custom callback handler.
 * Automatically traces all LLM calls, chain runs, and tool invocations.
 *
 * npm install arkvoid @langchain/core @langchain/openai langchain
 */

import { ArkvoidClient } from "arkvoid";
import type { ToolCallRecord } from "arkvoid";
import { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { Serialized } from "@langchain/core/load/serializable";
import type { LLMResult } from "@langchain/core/outputs";
import { ChatOpenAI } from "@langchain/openai";
import { AgentExecutor, createOpenAIToolsAgent } from "langchain/agents";
import { pull } from "langchain/hub";
import { TavilySearchResults } from "@langchain/community/tools/tavily_search";
import type { ChatPromptTemplate } from "@langchain/core/prompts";

// ─────────────────────────────────────────────
// ARKVOID LangChain Callback Handler
// ─────────────────────────────────────────────

export class ArkvoidLangChainHandler extends BaseCallbackHandler {
  name = "ArkvoidLangChainHandler";

  private client: ArkvoidClient;
  private agentSlug: string;
  private runStartTimes = new Map<string, number>();
  private pendingToolCalls = new Map<string, ToolCallRecord[]>();

  constructor(client: ArkvoidClient, agentSlug: string) {
    super();
    this.client = client;
    this.agentSlug = agentSlug;
  }

  // ── LLM Events ──────────────────────────────

  async handleLLMStart(
    llm: Serialized,
    _prompts: string[],
    runId: string
  ): Promise<void> {
    this.runStartTimes.set(runId, Date.now());
    this.pendingToolCalls.set(runId, []);
    console.log(`[ARKVOID] LLM started: ${llm.id?.join("/")} (${runId})`);
  }

  async handleLLMEnd(output: LLMResult, runId: string): Promise<void> {
    const startTime = this.runStartTimes.get(runId);
    const durationMs = startTime ? Date.now() - startTime : undefined;
    const toolCalls = this.pendingToolCalls.get(runId) ?? [];

    const generation = output.generations[0]?.[0];
    const modelName =
      (output.llmOutput?.model_name as string) ??
      (output.llmOutput?.modelName as string) ??
      "unknown";

    await this.client.trace({
      action: "llm_call",
      riskLevel: "low",
      agent: this.agentSlug,
      modelProvider: "openai",
      modelName,
      inputTokens: output.llmOutput?.tokenUsage?.promptTokens as number,
      outputTokens: output.llmOutput?.tokenUsage?.completionTokens as number,
      durationMs,
      ...(generation && { outputData: generation.text }),
      toolCalls,
      metadata: {
        source: "langchain",
        runId,
        generationCount: output.generations.length,
      },
      tags: ["langchain", "llm"],
    });

    this.runStartTimes.delete(runId);
    this.pendingToolCalls.delete(runId);
  }

  async handleLLMError(error: Error, runId: string): Promise<void> {
    const startTime = this.runStartTimes.get(runId);
    const durationMs = startTime ? Date.now() - startTime : undefined;

    await this.client.trace({
      action: "llm_error",
      riskLevel: "high",
      agent: this.agentSlug,
      durationMs,
      metadata: {
        source: "langchain",
        runId,
        error: error.message,
        success: false,
      },
      tags: ["langchain", "error"],
    });

    this.runStartTimes.delete(runId);
    this.pendingToolCalls.delete(runId);
  }

  // ── Tool Events ──────────────────────────────

  async handleToolStart(
    tool: Serialized,
    input: string,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    this.runStartTimes.set(runId, Date.now());
    console.log(
      `[ARKVOID] Tool started: ${tool.name ?? "unknown"} (${runId})`
    );

    // Register tool call under parent LLM run
    if (parentRunId && this.pendingToolCalls.has(parentRunId)) {
      const calls = this.pendingToolCalls.get(parentRunId)!;
      calls.push({
        toolName: (tool.name as string) ?? "unknown",
        callIndex: calls.length,
        input: { query: input },
        status: "success",
      });
    }
  }

  async handleToolEnd(
    output: string,
    runId: string,
    parentRunId?: string
  ): Promise<void> {
    const startTime = this.runStartTimes.get(runId);
    const durationMs = startTime ? Date.now() - startTime : undefined;

    // Update tool call record with output + latency
    if (parentRunId && this.pendingToolCalls.has(parentRunId)) {
      const calls = this.pendingToolCalls.get(parentRunId)!;
      const last = calls[calls.length - 1];
      if (last) {
        last.latencyMs = durationMs;
        last.output = { result: output.slice(0, 500) }; // truncate for preview
      }
    }

    this.runStartTimes.delete(runId);
  }

  async handleToolError(error: Error, runId: string): Promise<void> {
    await this.client.trace({
      action: "tool_error",
      riskLevel: "medium",
      agent: this.agentSlug,
      metadata: {
        source: "langchain",
        runId,
        error: error.message,
      },
      tags: ["langchain", "tool", "error"],
    });

    this.runStartTimes.delete(runId);
  }

  // ── Chain Events ─────────────────────────────

  async handleChainStart(
    _chain: Serialized,
    _inputs: Record<string, unknown>,
    runId: string
  ): Promise<void> {
    this.runStartTimes.set(runId, Date.now());
  }

  async handleChainEnd(
    _outputs: Record<string, unknown>,
    runId: string
  ): Promise<void> {
    this.runStartTimes.delete(runId);
  }
}

// ─────────────────────────────────────────────
// Usage: LangChain Agent with ARKVOID tracing
// ─────────────────────────────────────────────

const arkvoid = new ArkvoidClient({
  apiKey: process.env.ARKVOID_API_KEY!,
  agent: "research-agent",
});

const arkvoidHandler = new ArkvoidLangChainHandler(arkvoid, "research-agent");

const llm = new ChatOpenAI({
  model: "gpt-4o",
  temperature: 0,
  callbacks: [arkvoidHandler],
});

async function runResearchAgent(query: string) {
  const tools = [new TavilySearchResults({ maxResults: 3 })];

  const prompt = await pull<ChatPromptTemplate>(
    "hwchase17/openai-tools-agent"
  );

  const agent = await createOpenAIToolsAgent({ llm, tools, prompt });

  const executor = new AgentExecutor({
    agent,
    tools,
    callbacks: [arkvoidHandler],
    verbose: false,
  });

  console.log(`\nRunning research agent for: "${query}"`);

  const result = await executor.invoke({ input: query });

  return result.output;
}

// ─────────────────────────────────────────────
// Simple chain tracing (without agent)
// ─────────────────────────────────────────────

async function tracedChain(prompt: string) {
  const simpleLlm = new ChatOpenAI({
    model: "gpt-4o-mini",
    callbacks: [arkvoidHandler],
  });

  const result = await simpleLlm.invoke(prompt);
  return result.content;
}

// ─────────────────────────────────────────────
// Run
// ─────────────────────────────────────────────

(async () => {
  console.log("🔗 ARKVOID LangChain Integration Example\n");

  const answer = await tracedChain(
    "Explain the basics of reinforcement learning in 2 sentences."
  );
  console.log("LLM:", answer);
})();
